nexus/src/app/sagas/instance_create.rs

```diff
@@ -609,17 +607,40 @@ async fn sic_alloc_server(
     //   schedule instances that belong to a cluster on different failure
     //   domains. See https://github.com/oxidecomputer/omicron/issues/1705.
 
-    osagactx
+    // TODO: Fix these values. They're wrong now, but they let us move
+    // forward with plumbing.
+    let params = sagactx.saga_params::<Params>()?;
+    let hardware_threads = params.create_params.ncpus.0;
+    let rss_ram = params.create_params.memory;
+    let reservoir_ram = omicron_common::api::external::ByteCount::from(0);
+
+    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+    let resources = db::model::Resources::new(
+        hardware_threads.into(),
+        rss_ram.into(),
+        reservoir_ram.into(),
+    );
+
+    let resource = osagactx
         .nexus()
-        .random_sled_id()
+        .reserve_on_random_sled(
+            instance_id,
+            db::model::SledResourceKind::Instance,
+            resources,
+        )
         .await
-        .map_err(ActionError::action_failed)?
-        .ok_or_else(|| Error::ServiceUnavailable {
-            internal_message: String::from(
-                "no sleds available for new Instance",
-            ),
-        })
-        .map_err(ActionError::action_failed)
+        .map_err(ActionError::action_failed)?;
+    Ok(resource.sled_id)
+}
+
+async fn sic_alloc_server_undo(
+    sagactx: NexusActionContext,
+) -> Result<(), anyhow::Error> {
+    let osagactx = sagactx.user_data();
+    let instance_id = sagactx.lookup::<Uuid>("instance_id")?;
+
+    osagactx.nexus().delete_sled_reservation(instance_id).await?;
```

Comment:
FYI to @gjcolombo - "picking a sled" for an instance is now selected via `reserve_on_random_sled`, and removed via `delete_sled_reservation`.

It seems highly likely this will impact how instance migration will work -- we need to reserve on the destination, and eventually delete the reservation on the source sled.

Reply:
Thanks, this makes sense.

> we need to reserve on the destination

Is it possible to rearrange the parts of `sled_reservation_create` so that it's straightforward to add constraints to the search for sleds with available resources? In the migration case, where the external API currently allows a caller to select a destination sled explicitly, this would be "the sled ID is exactly X." More generally, I think we'll eventually want to be able to say, "this VM requires a sled with at least this software version" or "this instance must be in this particular rack" or even just "pick any compatible sled *except* this one." In a perfect world these would be easy to compose ("add these WHERE clauses to this query") but I'm not good enough with Diesel to assess how challenging that is.

> and eventually delete the reservation on the source sled

ACK. I already need to add a commit to the LM series that will deal with this sort of post-migration task, so this can be another piece of work that task has to do.

Reply:
@smklein and I talked this over offline and concluded that the easiest way to make progress for now is to do the following:

- Add a parameter to `sled_reservation_create` (either in this PR or in the live migration series, I'm not sure it matters which) that allows the caller to specify a sled, or set of sleds, to exclude from consideration. In the migration case, this is the instance's current sled.
- In my pending migration changes, change the migration saga to use `sled_reservation_create` instead of an explicitly-selected sled. (This would change the external API from "migrate this instance to sled X" to "migrate this instance to any eligible sled. cc @askfongjojo about this part--is this going to leave us with an unworkable migration API? Do we need to think about this more?)

If/when we need more constraints in the future, we can come up with a suitably elegant way to add them (e.g. a `SledSelectionConstraints` type with a fancy builder interface).

Reply:
I can add it to the migration series.